lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -1159,7 +1205,12 @@ def main():
         force_update=dict(default=False, type='bool'),
         log_driver=dict(default="json-file", type='str'),
         log_driver_options=dict(default={}, type='dict'),
-        publish=dict(default=[], type='list'),
+        publish=dict(default=[], type='list', elements='dict', options=dict(
+            published_port=dict(type='int', required=True),
+            target_port=dict(type='int', required=True),
+            protocol=dict(type='str', required=True),
```

Comment:
You can also add the `choices` here. Then Ansible will complain if the user specifies something different.